src/sessions/VirtualPageLoadTimer.ts

```diff
@@ -0,0 +1,245 @@
+        this.domMutationObserver.disconnect();
+
+        // Initialize timer objects and start observing
+        this.periodicCheckerId = setInterval(
+            this.checkLoadStatus,
+            this.config.routeChangeLoadCheckInterval
+        );
+        this.timeoutCheckerId = setTimeout(
+            this.declareTimeout,
+            this.config.routeChangeTimeout
+        );
+        this.domMutationObserver.observe(document, {
+            subtree: true,
+            childList: true,
+            attributes: true,
+            characterData: true
+        });
+
+        // Indicate page has not loaded, and carry over buffered requests.
+        this.isPageLoaded = false;
+        this.requestBuffer.forEach(this.moveItemsFromBuffer);
+        this.requestBuffer.clear();
+    }
+
+    protected patches(): MonkeyPatch[] {
+        return [
+            {
+                nodule: XMLHttpRequest.prototype,
+                name: 'send',
+                wrapper: this.sendWrapper
+            },
+            {
+                nodule: window,
+                name: 'fetch',
+                wrapper: this.fetchWrapper
+            }
+        ];
+    }
+
+    private sendWrapper = (): ((original: Send) => Send) => {
+        const self = this;
+
+        return (original: Send): Send => {
+            return function (this: XMLHttpRequest): void {
+                self.recordXhr(this);
+                this.addEventListener('loadend', self.endTracking);
+                return original.apply(this, arguments);
+            };
+        };
+    };
+
+    private recordXhr(item: XMLHttpRequest) {
+        const page = this.pageManager.getPage();
+        if (page && this.isPageLoaded === false) {
+            this.ongoingRequests.add(item);
+        } else {
+            this.requestBuffer.add(item);
+        }
+    }
+
+    private removeXhr(item: XMLHttpRequest, currTime: number) {
+        const page = this.pageManager.getPage();
+        if (page && this.ongoingRequests.has(item)) {
+            this.ongoingRequests.delete(item);
+            this.latestEndTime = currTime;
+        } else if (this.requestBuffer.has(item)) {
+            this.requestBuffer.delete(item);
+        }
+    }
+
+    /**
+     * Removes the current event from either requestBuffer or ongoingRequests set.
+     * @param event
+     */
+    private endTracking = (e: Event) => {
+        const currTime = Date.now();
+        const xhr: XMLHttpRequest = e.target as XMLHttpRequest;
+        xhr.removeEventListener('loadend', this.endTracking);
+        this.removeXhr(xhr, currTime);
+    };
+
+    private fetch = (
+        original: Fetch,
+        thisArg: Fetch,
+        argsArray: IArguments,
+        input: RequestInfo,
+        init?: RequestInit
+    ): Promise<Response> => {
+        const self = this;
+        return original
+            .apply(thisArg, argsArray)
+            .catch((error) => {
+                // Silently swallowing error
+            })
```

Comment:
issue: If we catch the error without re-throwing it, the error will not be visible to the application. This changes the application's behavior.

Reply:
Ack, will reflect in next revision. Adjusted unit test cases accordingly.